### PR TITLE
Base structure for the new charm store repository implementation.

### PR DESCRIPTION
**Pull request by frankban**
Rename the existing one to LegacyCharmStore.
Also implement initializers for the new charm store and the
local repository.

### CODE REVIEW

File: charmrepo/charmstore.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	// the user visits a web page to authenticate themselves.
+	// If nil, a default function that returns an error will be used.
+	VisitWebPage func(url *url.URL) error
+
+	// CacheDir holds the charm cache directory path where to store retrieved
+	// charms.
+	CacheDir string
+}
+
+// NewCharmStore creates and return a charm store repository.
```
Comment (bac) [inline charmrepo/charmstore.go]: typo: returns

File: charmrepo/legacy.go [hunk 2]
```diff
@@ -23,21 +23,22 @@ import (
 // CacheDir stores the charm cache directory path.
 var CacheDir string
 
-// CharmStore is a repository Interface that provides access to the public juju charm store.
-type CharmStore struct {
+// LegacyCharmStore is a repository Interface that provides access to the
+// public Juju charm store.
```
Comment (bac) [inline charmrepo/legacy.go]: Is "public" the correct description when this lands?

File: charmrepo/repo.go [hunk 3]
```diff
@@ -18,8 +17,14 @@ var logger = loggo.GetLogger("juju.charm.charmrepo")
...
 // Interface represents a charm repository (a collection of charms).
 type Interface interface {
+	// Get returns the charm referenced by curl.
 	Get(curl *charm.URL) (charm.Charm, error)
+
+	// Latest returns the latest revision of the charms referenced by curls,
+	// regardless of the revision set on each curl.
 	Latest(curls ...*charm.URL) ([]CharmRevision, error)
+
+	// Resolve canonicalizes charm URLs any implied series in the reference.
```
Comment (bac) [inline charmrepo/repo.go]: This sentence is missing a word or two ... I just don't know what they are.

**Comment by bac** [comment]
Code :+1: with some niggling comments.

**Comment by mhilton** [comment]
:+1: LGTM, with @bac's comments.

File: charmrepo/charmstore.go (hunk 1 above)
Comment (frankban) [inline charmrepo/charmstore.go]: Done.

File: charmrepo/legacy.go (hunk 2 above)
Comment (frankban) [inline charmrepo/legacy.go]: Done.

File: charmrepo/repo.go (hunk 3 above)
Comment (frankban) [inline charmrepo/repo.go]: Done.

**Pull request merged**